crates/common/consensus/src/deneb/beacon_state.rs

```diff
@@ -1312,6 +1313,78 @@ impl BeaconState {
         }
         Ok(())
     }
+
+    pub fn process_eth1_data(&mut self, body: BeaconBlockBody) -> anyhow::Result<()> {
+        self.eth1_data_votes
+            .push(body.eth1_data.clone())
+            .map_err(|err| anyhow!("Can't push eth1_data {err:?}"))?;
+
+        let count = self
+            .eth1_data_votes
+            .iter()
+            .filter(|data| **data == body.eth1_data)
+            .count() as u64;
+
+        if count * 2 > (EPOCHS_PER_ETH1_VOTING_PERIOD * SLOTS_PER_EPOCH) {
+            self.eth1_data = body.eth1_data;
+        }
+
+        Ok(())
+    }
+
+    pub fn process_randao_mixes_reset(&mut self) -> anyhow::Result<()> {
+        let current_epoch = self.get_current_epoch();
+        let next_epoch = current_epoch + 1;
+        // Set randao mix
+        self.randao_mixes[(next_epoch % EPOCHS_PER_HISTORICAL_VECTOR) as usize] =
+            self.get_randao_mix(current_epoch);
+        Ok(())
+    }
+
+    pub fn process_slashings_reset(&mut self) -> anyhow::Result<()> {
+        let next_epoch = self.get_current_epoch() + 1;
+        // Reset slashings
+        self.slashings[(next_epoch % EPOCHS_PER_SLASHINGS_VECTOR) as usize] = 0;
+        Ok(())
+    }
+
+    pub fn process_slashings(&mut self) -> anyhow::Result<()> {
+        let epoch = self.get_current_epoch();
+        let total_balance = self.get_total_active_balance();
+        let adjusted_total_slashing_balance = (self.slashings.iter().sum::<u64>()
+            * PROPORTIONAL_SLASHING_MULTIPLIER_BELLATRIX)
+            .min(total_balance);
+
+        for index in 0..self.validators.len() {
+            if self.validators[index].slashed
+                && epoch + EPOCHS_PER_SLASHINGS_VECTOR / 2
+                    == self.validators[index].withdrawable_epoch
+            {
+                let increment = EFFECTIVE_BALANCE_INCREMENT;
+                let penalty_numerator = self.validators[index].effective_balance / increment
+                    * adjusted_total_slashing_balance;
+                let penalty = penalty_numerator / total_balance * increment;
+
+                self.decrease_balance(index as u64, penalty);
+            }
+        }
```

Comment:
```suggestion
        for (index, validator) in self.validators.iter().enumerate() {
            if validator.slashed && epoch + EPOCHS_PER_SLASHINGS_VECTOR / 2 == validator.withdrawable_epoch
            {
                let increment = EFFECTIVE_BALANCE_INCREMENT;
                let penalty_numerator = validator.effective_balance / increment * adjusted_total_slashing_balance;
                let penalty = penalty_numerator / total_balance * increment;

                self.decrease_balance(index as u64, penalty);
            }
        }
```

you can use enumerate to avoid some of that indexing